File: problems/leetcode/0100/113.cpp
// 113. Path Sum II
// https://leetcode.com/problems/path-sum-ii/description/

using namespace std;

#include <vector>

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  vector<vector<int>> paths;

  void dfs(TreeNode *node, vector<int> &path, int prevSum, int targetSum) {
    if (!node) {
      return;
    }

    path.push_back(node->val);

    int sum = prevSum + node->val;
    if (sum == targetSum && !node->left && !node->right) {
      paths.push_back(path);
    }

    dfs(node->left, path, sum, targetSum);
    dfs(node->right, path, sum, targetSum);

    path.pop_back();
  }

  vector<vector<int>> pathSum(TreeNode *root, int targetSum) {
    vector<int> path;
    dfs(root, path, 0, targetSum);
    return paths;
  }
};

File: problems/leetcode/0100/113/113.py
# 113. Path Sum II
# https://leetcode.com/problems/path-sum-ii/description/

from typing import *

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        result, currentPath = [], []
        self.traverse(root, targetSum, currentPath, 0, result)
        return result

    def traverse(self, node: Optional[TreeNode], targetSum: int, currentPath: List[int], currentSum: int, result: List[List[int]]) -> None:
        if not node:
            return

        currentPath.append(node.val)

        if not node.left and not node.right and targetSum == currentSum + node.val:
            result.append([n for n in currentPath])
            currentPath.pop()
            return

        for child in [node.left, node.right]:
            if child:
                self.traverse(child, targetSum, currentPath, currentSum + node.val, result)

        currentPath.pop()
